src/bloom.rs

```diff
@@ -0,0 +1,133 @@
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+
+pub trait BitSlice {
+    fn get_bit(&self, idx: usize) -> bool;
+    fn bit_len(&self) -> usize;
+}
+
+pub trait BitSliceMut {
+    fn set_bit(&mut self, idx: usize, val: bool);
+}
+
+impl<'a, T: AsRef<[u8]>> BitSlice for T {
+    fn get_bit(&self, idx: usize) -> bool {
+        let pos = idx / 8;
+        let offset = idx % 8;
+        (self.as_ref()[pos] & (1 << offset)) != 0
+    }
+
+    fn bit_len(&self) -> usize {
+        self.as_ref().len() * 8
+    }
+}
+
+impl<'a, T: AsMut<[u8]>> BitSliceMut for T {
+    fn set_bit(&mut self, idx: usize, val: bool) {
+        let pos = idx / 8;
+        let offset = idx % 8;
+        if val {
+            self.as_mut()[pos] |= 1 << offset;
+        } else {
+            self.as_mut()[pos] &= !(1 << offset);
+        }
+    }
+}
+
+/// Bloom implements bloom filter functionalities over
+/// a bit-slice of data.
+pub struct Bloom<'a> {
+    filter: &'a [u8],
+    k: u32,
+}
+
+impl<'a> Bloom<'a> {
+    /// Create a bloom filter from a byte slice
+    pub fn new(buf: &'a [u8]) -> Self {
+        let filter = &buf[..buf.len() - 4];
+        let k = (&buf[buf.len() - 4..]).get_u32();
+        Self { filter, k }
+    }
+
+    /// Get bloom filter bits per key from entries count and FPR
+    pub fn bloom_bits_per_key(entries: usize, false_positive_rate: f64) -> usize {
+        let size =
+            -1.0 * (entries as f64) * false_positive_rate.ln() / (0.69314718056 as f64).powi(2);
+        let locs = 0.69314718056 * size / (entries as f64);
+        locs as usize
+    }
+
+    /// Build bloom filter from key hashes
+    pub fn build_from_key_hashes(keys: &[u32], bits_per_key: usize) -> Bytes {
+        let k = ((bits_per_key as f64) * 0.69) as u32;
+        let k = k.min(30).max(1);
+        let nbits = (keys.len() * bits_per_key).max(64);
+        let nbytes = (nbits + 7) >> 3;
+        // nbits is always multiplication of 8
+        let nbits = nbytes << 3;
+        let mut filter = Vec::with_capacity(nbytes);
+        filter.resize(nbytes, 0);
+        for h in keys {
+            let mut h = *h;
+            let delta = (h >> 17) | (h << 15);
+            for _ in 0..k {
+                let bit_pos = (h as usize) % nbits;
+                filter.set_bit(bit_pos, true);
+                h = h.wrapping_add(delta);
+            }
+        }
+        let mut buf = BytesMut::new();
+        buf.put_slice(filter.as_slice());
+        buf.put_u32(k);
```

Comment:
It seems in `[1, 30]`.

Reply:
k is set to be in `[1, 30]` at the beginning of this func by `let k = k.min(30).max(1);`